### PR TITLE
Update Layout.py

### PR DESCRIPTION
**Pull request by mg979**
Hi there, I made several modifications to your very good plugin. Sorry I'm not practical with GitHub (first time I use). Also first Python script for that matter.

- line 8, I get errors sometimes, it doesn't find the path if os functions are in the global section(I don't know why)

- line 56, added refresh_layout, called after each split/reload of layout, because new panes aren't drawn immediately

- altered load/save to json to remember and restore active pane after reload (I want to change it again to make restoring optional, and even filename storing, so that one can cycle layouts without even changing files)

- load/save layout input panels now get the focus

- the resize_pane stuff didn't work well for panes that were in the middle, vertically or horizontally. Now panes in the middle can be resized in all directions. Normal resizing happens along the same axis, no matter the position of the pane, but if the "alt" argument is true (another keybinding), resizing uses another axis of the same frame, if in the middle, or no change if it's an external one. Probably not 100% perfect code-wise but it seems to work.

I'm using these key bindings:

        { "keys": ["alt+h"],               "command": "layout_resize", "args": {"direction": "left", "step": 1} },
        { "keys": ["alt+l"],               "command": "layout_resize", "args": {"direction": "right", "step": 1} },
        { "keys": ["alt+k"],               "command": "layout_resize", "args": {"direction": "up", "step": 1} },
        { "keys": ["alt+j"],               "command": "layout_resize", "args": {"direction": "down", "step": 1} },
        { "keys": ["ctrl+alt+h"],          "command": "layout_resize", "args": {"direction": "left", "step": 1, "alt": true} },
        { "keys": ["ctrl+alt+l"],          "command": "layout_resize", "args": {"direction": "right", "step": 1, "alt": true} },
        { "keys": ["ctrl+alt+k"],          "command": "layout_resize", "args": {"direction": "up", "step": 1, "alt": true} },
        { "keys": ["ctrl+alt+j"],          "command": "layout_resize", "args": {"direction": "down", "step": 1, "alt": true} },
        { "keys": ["ctrl+alt+shift+k"],    "command": "resize_pane", "args": {"direction": "out" }},
        { "keys": ["ctrl+alt+shift+j"],    "command": "resize_pane", "args": {"direction": "in" }},

I renamed some of the functions (to avoid overlap with Origami and similar plugins). "resize_pane" is from Pane Resizer, for radial pane resizing.

### CODE REVIEW

**Comment by mg979** [comment]
I updated with the changes you suggested (forgot self.window.show_input_panel instead of sublime.active_window().show_input_panel, I will add it in the next). Sorry for the imperfections of the code, I'm very new to python and sublime text. I used Anaconda autoformat for pep8 errors and it cuts lines to 80, I'll add an exception for long lines.

The load/save layout feature still isn't as I'd like it to be, I'll change things to something like this:

- modes for 'load from file':
    - layout: load layout and ignore saved layout tabs
    - add: load saved layout tabs but keep current too, this is current behavior
    - replace: load layout and discard everything that wasn't in it (unless dirty)

- load from file input: I'll add a quick panel to choose a file from, and also to select the mode above

File: Layout.py [hunk 1]
```diff
@@ -2,16 +2,16 @@
...
 
 PLUGIN_NAME = 'Layout'
-PLUGIN_PATH = os.path.join(sublime.packages_path(), PLUGIN_NAME)
-LAYOUTS_PATH = os.path.join(PLUGIN_PATH, 'layouts')
 
 X_MIN, Y_MIN, X_MAX, Y_MAX = (0, 1, 2, 3)
 
 
 def plugin_loaded():
+    PLUGIN_PATH = os.path.join(sublime.packages_path(), PLUGIN_NAME)
```
Comment (mg979) [inline Layout.py]: I keep getting errors for loading layouts after closing and reopening Sublime, if these os.path functions are in the global section, so I moved them in defs

File: Layout.py [hunk 2]
```diff
@@ -2,16 +2,16 @@
...
+    LAYOUTS_PATH = os.path.join(PLUGIN_PATH, 'layouts')
     if not os.path.isdir(LAYOUTS_PATH):
```
Comment (mg979) [inline Layout.py]: I keep getting errors for loading layouts after closing and reopening Sublime, if these os.path functions are in the global section, so I moved them in defs

File: Layout.py [hunk 3]
```diff
@@ -2,16 +2,16 @@
...
 PLUGIN_NAME = 'Layout'
-PLUGIN_PATH = os.path.join(sublime.packages_path(), PLUGIN_NAME)
-LAYOUTS_PATH = os.path.join(PLUGIN_PATH, 'layouts')
 
 X_MIN, Y_MIN, X_MAX, Y_MAX = (0, 1, 2, 3)
 
 
 def plugin_loaded():
+    PLUGIN_PATH = os.path.join(sublime.packages_path(), PLUGIN_NAME)
+    LAYOUTS_PATH = os.path.join(PLUGIN_PATH, 'layouts')
```
Comment (mg979) [inline Layout.py]: I keep getting errors for loading layouts after closing and reopening Sublime, if these os.path functions are in the global section, so I moved them in defs

File: Layout.py [hunk 4]
```diff
@@ -35,6 +36,8 @@ def current_group(self):
 
     @classmethod
     def get_layout_path(cls, filename):
+        PLUGIN_PATH = os.path.join(sublime.packages_path(), PLUGIN_NAME)
+        LAYOUTS_PATH = os.path.join(PLUGIN_PATH, 'layouts')
```
Comment (mg979) [inline Layout.py]: same here

File: Layout.py [hunk 5]
```diff
@@ -47,30 +50,60 @@ def fixed_set_layout(self, layout):
...
+
+        for group in range(len(layout['cells'])):
+            views = layout['views_in_group'][group]
+            for view in views:
+                if view['is_active']:
+                    active = view
+                    views.remove(view)
+                    break
+            views.append(active)
+
```
Comment (mg979) [inline Layout.py]: this pulling out/back in maybe isn't necessary, but for now I'm keeping it

File: Layout.py [hunk 6]
```diff
@@ -326,6 +416,23 @@ def get_options(self, command):
...
+    We only want to get the quick panel view, so we use a flag
+    View saved as an attribute of the global variable sublime
+    so it can be accesed from your plugin or anywhere"""
+
+    def on_activated(self, view):
+        if sublime.capturingQuickPanelView is True:
+            sublime.capturingQuickPanelView = False
+            sublime.quickPanelView = view
+
+
```
Comment (mg979) [inline Layout.py]: this is for the layout preview in quick panel when loading from file

File: Layout.py [hunk 7]
```diff
@@ -421,63 +529,146 @@ def do_run(self):
...
+                    caption='Load from Layout File',
+                    initial_text='',
+                    on_done=self.load_layout_from_file,
+                    on_change=None,
+                    on_cancel=None
+                )
+                """
+
+        self.window.show_quick_panel(modes, on_done)
+
```
Comment (mg979) [inline Layout.py]: This is the quick panel for choosing saved layouts, with preview

File: Layout.py [hunk 8]
```diff
@@ -163,40 +233,60 @@ def resize_pane(self, group, direction, step):
...
+                    rows[y2] += step
+            elif y1 > 0 and y2 < max_y:
+                if rows[y2] - rows[y2 - 1] > MIN_GAP:
+                    rows[y1] += step
+            else:
+                gy = 1 if y2 + 1 > max_y else y2 + 1
+                if y2 < max_y and rows[gy] - rows[y2] > MIN_GAP:
+                    rows[y2] += step
+                if y2 == max_y and rows[y2] - rows[y1] > MIN_GAP:
+                    rows[y1] += step
```
Comment (mg979) [inline Layout.py]: This is how you can resize frames in the middle with different keybindings and alt=true argument

**Comment by mg979** [comment]
This is how you can resize frames in the middle with different keybindings and alt=true argument

**Comment by loggerhead** [comment]
Sorry for the delay. The code still needs improve, some little things like:

* redundancy empty lines
* removing when an iteration
* the format of docstring

And putting the explanation of some function in the comments (if you need explain it in picture, you should consider using ASCII graphs), people who wants to modify the codes would be easy understand the purpose of the function.

You can review some other projects to see how do they do. Also, feel free to propose any problem here :)

**Comment by mg979** [comment]
Ok I will remove some of the most recent changes that I made, I will only include some basic fixes (refresh layout and resize panes in all directions), I will work on bigger changes in a new branch in my fork to avoid messing up stuff.

I will add better descriptions to all new stuff, not sure what you mean by "removing when an iteration", do you refer to that "focus on active tab" stuff? I think I'll remove it entirely and handle it differently, but in the new branch, for now I just remove it.

**Review by mg979 — COMMENTED** [review]
Please select "Changes from all commits" when viewing the last commit, there are far less changes compared to your master than in my last commit.

**Pull request closed without merging**